WIP: General Improvements Concerning `pyproject.py`

eeriemyxi opened the pull request:
# Technical Details
- Move implementation to `src` folder.
- Move dependency index to `pyproject.py`, delete `requirements.txt` then. Also pin-point working dependency version in the index.
- Add `mswaa` executable in the `pyproject.py` so it gets installed when the package is installed.
- Update `README.md` with newer and simpler installation instructions.
- Get rid of shell scripts like `setup.sh` and `main.sh` since they're useless now.
  - There is a section called Development/Contributing in the README which relies on `uv` to do the development setup (which I assume is the purpose of `setup.sh`). If you'd like, I can bring the `setup.sh` back if you don't like that workflow (hypothetically, it should still work with the new structure.)

This pull request tries to resolve issues with the current installation method which is somewhat messy:
- It relies on shell scripts `main.sh` to create an executable for the project, even though this has been standardized in Python build systems for a very long time already. See `[project.scripts]` for reference.
- `requirements.py` file doesn't have the dependency versions locked, which can lead to unusual bugs as those projects mature. The need for `requirements.py` has also been more thoroughly standardized in the `pyproject.py` spec. See `[project.dependencies]` in `pyproject.py`. Thus removed `requirements.py` and moved its purpose to `pyproject.py`.

I have updated the `README.md` with the new instructions. If you'd like, you can test the new installation instruction before merging:

```python
cd $(mktemp -d)
git clone --depth 1 https://github.com/eeriemyxi-contributions/mouse-scroll-wheel-acceleration-userspace mswaa 
git checkout 5fa95ffcfef914a1ce5b9bcf833c8b49c3393b69
cd mswaa
python -m venv .venv
chmod +x .venv/bin/activate
.venv/bin/activate
pip3 install .
mswaa --help
sleep 3; ./install-daemon.py
deactivate
```

eeriemyxi committed:
refactor: several refactors (read commit description)

- Move implementation to `src` folder.
- Move dependency index to `pyproject.py`, delete `requirements.txt` then. Also
  pin-point working dependency version in the index.
- Add `mswaa` executable in the `pyproject.py` so it gets installed when the
  package is installed.
- Update `README.md` with newer and simpler installation instructions.
- Get rid of shell scripts like `setup.sh` and `main.sh` since they're useless
now.

albertz commented:
@valmikikothare @xythobuz @blueyed What do you think about this? (Btw, I can add you to this project, so that you can properly review, if you want.)

eeriemyxi committed:
fix: installer should exit if server_bin is None

eeriemyxi committed:
fix: `deactivate` -> `exit`

valmikikothare commented:
@eeriemyxi How does the script installed through the pyproject.toml file handle virtual environment activation? The point of `setup.sh` and `main.sh` were to create an isolated virtual environment where the necessary dependencies would be installed and to activate that environment whenever the daemon was started/restarted, respectively. I made that choice specifically to avoid `pip install --user` so as not to interfere with any of the user's other projects. Is there a way to do this with pyproject.toml?

@albertz That would be great, thanks

valmikikothare commented:
To follow up, the rest of these changes mostly make sense and I agree that the need for bash scripts is a bit hacky. 
I'm not sure I understand the need for uv in this case though, as the development dependencies are the same as the usage dependencies and are installed alongside the package to begin with. If you want to ensure exact version requirements for development purposes, why not use `requirements.txt`, given this is a fairly small scale project (don't get me wrong, `uv` seems great :)
In addition, the dependency version requirements in `pyproject.toml` seem a bit restrictive, are you certain that older versions would not work? At least some of them seem to be the newest versions available. I believe it's generally best practice to keep the version requirements in `pyproject.toml` as nonrestricitve as possible and to provide specific dependency versioning for a known working environment setup in the `requirements.txt` (or `uv.lock`) file.

eeriemyxi commented:
> @eeriemyxi How does the script installed through the pyproject.toml file handle virtual environment activation? The point of `setup.sh` and `main.sh` were to create an isolated virtual environment where the necessary dependencies would be installed and to activate that environment whenever the daemon was started/restarted, respectively. I made that choice specifically to avoid `pip install --user` so as not to interfere with any of the user's other projects. Is there a way to do this with pyproject.toml?
> 
> @albertz That would be great, thanks

Hi, the new installation instructions don't take that into account, mostly due to my misjudgement.

Since `pyproject.py` does most of the heavy-lifting, this lets people who don't want or can't use virtual environments in their system to also make use of the package. Instead virtual environments can be documented as an additional feature that we can include in the installation instructions by default too.

Here are two solutions:
```bash
uv tool install git+https://github.com/albertz/mouse-scroll-wheel-acceleration-userspace
```
`uv tool` handles the management of virtual environments for us. The `install-daemon.py` script, in my commit, only expects a `mswaa` to be in the path, which `uv tool` will manage for us (under an isolated virtual environment.)

Here is the second solution:
```bash
pipx install git+https://github.com/albertz/mouse-scroll-wheel-acceleration-userspace
```

One expects `uv` to be installed, the other expects `pipx` to be installed. If you use `uv`, it will also handle Python runtime versions for you in the shadow as a bonus.

Depending on which tool maintainers of this project like, I can happily include the installation instructions for the said tool (uv or pipx) into the README. For both of these a simple `pip install` command should do for installation.

eeriemyxi commented:
> To follow up, the rest of these changes mostly make sense and I agree that the need for bash scripts is a bit hacky. I'm not sure I understand the need for uv in this case though, as the development dependencies are the same as the usage dependencies and are installed alongside the package to begin with. If you want to ensure exact version requirements for development purposes, why not use `requirements.txt`, given this is a fairly small scale project (don't get me wrong, `uv` seems great :) In addition, the dependency version requirements in `pyproject.toml` seem a bit restrictive, are you certain that older versions would not work? At least some of them seem to be the newest versions available. I believe it's generally best practice to keep the version requirements in `pyproject.toml` as nonrestricitve as possible and to provide specific dependency versioning for a known working environment setup in the `requirements.txt` (or `uv.lock`) file.

> If you want to ensure exact version requirements for development purposes, why not use requirements.txt, given this is a fairly small scale project (don't get me wrong, uv seems great :)

I think the maintainers will appreciate having a lot of more things to be in a coherent place. Since this project was already using `pyproject.toml` for configuring tools like `black`, so I don't see why not also use it for maintaining dependency especially since it has been standardized by the Python community now. This also simplifies the installation, because if you know what you're doing, you can just do `pip install .` to install the package and `pyproject.toml` takes care of everything.

> I believe it's generally best practice to keep the version requirements in pyproject.toml as nonrestricitve as possible and to provide specific dependency versioning for a known working environment setup in the requirements.txt (or uv.lock) file.

I agree with this, but I didn't really have much information to work with. The `requirements.txt` was like this:
```
better_exchook
pynput
appdirs
```

Without the versions specified, this requirements file tells the package manager to grab the latest package. I don't find my system qualified to judge whether or not a particular old version of these packages will work out for everyone. So I decided to go with the latest versions. If you have certain versions in mind, please let me know and I'll update the file.

albertz commented:
Despite the discussion on virtual environment, I was wondering about the abbreviation "mswaa". I don't understand this abbreviation. Why "aa" at the end? And I also don't like it too much. Can't we keep the full name? Or maybe shorter name but still descriptive enough? I think you won't type this too often, and when you look through your Python packages, it's good when you can easily identify the package by name. Opinions?

eeriemyxi commented:
> Despite the discussion on virtual environment, I was wondering about the abbreviation "mswaa". I don't understand this abbreviation. Why "aa" at the end? And I also don't like it too much. Can't we keep the full name? Or maybe shorter name but still descriptive enough? I think you won't type this too often, and when you look through your Python packages, it's good when you can easily identify the package by name. Opinions?

Another error on my part. It should've been `mswau`. If you don't like that either, let me know a name you do like, I'll update the changes accordingly.

PS: **m**ouse-**s**croll-**w**heel-**a**cceleration-**u**serspace -> mswau

valmikikothare commented:
> I think the maintainers will appreciate having a lot of more things to be in a coherent place. Since this project was already using pyproject.toml for configuring tools like black, so I don't see why not also use it for maintaining dependency especially since it has been standardized by the Python community now. This also simplifies the installation, because if you know what you're doing, you can just do pip install . to install the package and pyproject.toml takes care of everything.

I agree that the dependencies should be included in the `pyproject.toml`, I don't mean to suggest the `requirements.txt` should replace this for typical user installation. In my mind, there would be two installation methods:

1. `pip install --user .` (or `pip install --user git+https://github.com/albertz/mouse-scroll-wheel-acceleration-userspace.git`)
2. `./setup.sh` (or just including the contents of `setup.sh` in the `README` and instructing users to do that, namely the creation of a virutal environmnt and installing the requirements in that environment). We may find it useful to have a requirements.txt as a snapshot of a working environment, but it is not strictly necessary. However, we would need a shell script (i.e. `main.sh`) or similar to activate the environment before calling the daemon script, unless calling `mswau` (or whatever the command ends up being) automatically sources the environment before executing, regardless of the currently active environment. 

The explicit virtual environment creation, shell script use, and `requirements.txt` inclusion can be replaced by `uv` or `pipx` if you desire and if it significantly eases the installation process for a virtual environment setup. I just hesitate to require the use of a broad-scope, all-purpose tool for an otherwise extremely lightweight package, although I will defer to your judgement and that of the other maintainers.

valmikikothare commented:
@albertz I think having an abbreviated name for the command line is not a bad idea, though something like `scroll-accel` might be more descriptive. That being said, it might be worth shortening the name of the project as a whole to something simpler like Scroll Wheel Acceleration and including the specifics of the userspace installation in the `README`, but that is entirely up to you.

eeriemyxi commented:
> 2. ./setup.sh (or just including the contents of setup.sh in the README and instructing users to do that, namely the creation of a virutal environmnt and installing the requirements in that environment). We may find it useful to have a requirements.txt as a snapshot of a working environment, but it is not strictly necessary. However, we would need a shell script (i.e. main.sh) or similar to activate the environment before calling the daemon script, unless calling mswau (or whatever the command ends up being) automatically sources the environment before executing, regardless of the currently active environment.

I do like giving people choices, so I don't find any issue in providing an alternative and more portable solution. But would you agree that this version of the installation is complicated enough to be expected to be done by power users? If you do, I think I have a better solution to this which is somewhat similar.

The solution is to tell them to put a bash script called `mswaa` in their PATH, such that `mswaa` is an executable command. This `mswaa` file will resolve the virtual environment and launch the tool under its scope. 

Now for the `mswaa` executable, we can also provide a helper script at `/extras/install.py` which will essentially put an executable bash script at `$HOME/.local/bin` called `mswaa`. The install script will do one thing, which is making a persistent virtual environment for the project somewhere and embed it into the bash script. Then we document this helper in the README.

> [!NOTE]
> `mswaa` is a placeholder for the eventual name of the project.

albertz commented:
I wonder a bit whether there needs to be a binary at all in `$PATH`. For debugging / playing around with options, I just directly started the script from the directory. (If this is a PIP package now, then maybe you would do `python -m scroll_accel` or so.) Or otherwise I started it as a daemon. But I never really saw the need for a binary in `$PATH`. I'm just wondering. If you think that's useful, then let's go ahead. But do you agree that the usage as a daemon is probably more common? Maybe the daemon-installer script should actually be an executable in `$PATH`? Or the main executable should also offer an option to install it as a daemon, i.e. to call the daemon-installer script?

eeriemyxi commented:
I think the executable is useful, because without it the daemon installer will have to guess which Python runtime the user is talking about. With the executable, the installer only looks for an executable in the $PATH, so the user can customize which Python runtime to use.

> Or the main executable should also offer an option to install it as a daemon, i.e. to call the daemon-installer script?

I think something like `mswau update-daemon` would indeed be cool. But I'd like more opinions on it.

valmikikothare commented:
> With the executable, the installer only looks for an executable in the $PATH, so the user can customize which Python runtime to use.

If this is the case, does it not install the package/dependencies into the currently active environment, and if so, does it not use the Python runtime and packages associated with that environment? If that's the case, then your solution is fine because the user can handle the virtual environment creation on their own and just install the package into whatever environment they choose, or just install it into the `--user` space. Additionally, if this is the case, there should be no need for `uv` or `pipx` at all. 

> But do you agree that the usage as a daemon is probably more common? Maybe the daemon-installer script should actually be an executable in $PATH? Or the main executable should also offer an option to install it as a daemon, i.e. to call the daemon-installer script?

I think if we want to use the executable (which I agree we should), it should encompass testing the user's configuration (essentially by calling the `main()` function in `main.py`), updating the config file with the desired values, and installing/uninstalling the daemon. I.e. as a user you should never need to call anything other than the installed command executable (e.g. `mswau`, `scroll-accel`, etc.). 

Ideal usage *(Note: I will use `mswau` as the command, knowing this will likely change)*:

1. [Optional] Create and activate your environment, e.g.:
    ```bash
    python -m venv venv
    source venv/bin/activate
    ```
    or 
    ```bash
    conda create -n mswau python
    conda activate mswau
    ```
    or any other python environment manager

2. Install the package:
    ```bash
    pip install [--user] git+https://github.com/albertz/mouse-scroll-wheel-acceleration-userspace.git
    ```

3. Test your configs:
```bash
mswau test --multiplier 2.0 --exp 0.5 --threshold 2.0
```

4. Install the daemon with your desired configs:
```bash
mswau install_daemon --multiplier ...
```

5. Change your config values:
```bash
mswau update_daemon --multiplier ...
```

6. If hoping to uninstall:
```bash
mswau uninstall_daemon
```

Let me know if you guys agree

valmikikothare commented:
Also something I just noticed, the main logic has been moved to `__init__.py`. Generally I don't think this is good practice, there should be an explicit, named file (e.g. `scroll_accelerator.py`, `main.py`, etc.) with all the logic of the script, including the command line arguments, and `__init__.py` should only handle import logic for the necessary functions/modules from the directory in which it resides.

eeriemyxi commented:
> Also something I just noticed, the main logic has been moved to __init__.py. Generally I don't think this is good practice, there should be an explicit, named file (e.g. scroll_accelerator.py, main.py, etc.) with all the logic of the script, including the command line arguments, and __init__.py should only handle import logic for the necessary functions/modules from the directory in which it resides.

Yesterday I cleared one of my misconceptions about Python that you don't need `__main__.py` to make a module executable via `py -m`. But I was wrong. I'll be moving over this bit of code from `__init__.py` to `__main__.py`:

```python
if __name__ == "__main__":
    better_exchook.install()
    try:
        main()
    except KeyboardInterrupt:
        print("KeyboardInterrupt")
        sys.exit(1)
```

Including the relevant imports of course. Now `__init__.py` only contains the implementation, with the runtime shifted to `__main__.py`.

> Additionally, if this is the case, there should be no need for uv or pipx at all.

Yes, `uv` and `pipx` are _technically_ not needed, but they do provide significant convenience. They automatically handle allocation and deallocation of virtual environments of installed packages, so you don't need to care about it yourself at all. Which is useful when you just want to use the software. I think the recommended instructions in the README using one of them makes sense.

In my opinion package handling in Python is flawed by design: you shouldn't need virtual environments at all if you have a "package manager" (pip). So if that's the case I think it makes sense to recommend tools to remedy this issue.

eeriemyxi committed:
fix: move the runtime code to __main__.py from __init__.py

eeriemyxi committed:
fix: rm useless comment

I just noticed that the error is from our side (`sys.exit(1)`) not `better_exchook`.

valmikikothare commented:
> Yesterday I cleared one of my misconceptions about Python that you don't need __main__.py to make a module executable via py -m. But I was wrong. I'll be moving over this bit of code from __init__.py to __main__.py

> Including the relevant imports of course. Now __init__.py only contains the implementation, with the runtime shifted to __main__.py

This change still misses the point of removing implementation from `__init__.py`. See https://packaging.python.org/en/latest/guides/creating-command-line-tools/ for a good example (CLI logic in `cli.py`, empty `__init__.py`, code calling the CLI in `__main__.py`). Happy to make this change myself if you'd like. 

> In my opinion package handling in Python is flawed by design: you shouldn't need virtual environments at all if you have a "package manager" (pip). So if that's the case I think it makes sense to recommend tools to remedy this issue.

I don't think this is a widely held opinion, virtual environments offer great reproducibility and provide essential isolation of your projects from each other should they require different dependencies. Environment management is widely used at various levels of isolation (`venv` to Docker). All that said, I don't have a problem with `uv` as it provides that very isolation, and will defer to your implementation so long as it is optional and can exist alongside a manual environment creation. 

@eeriemyxi please ping me when you have a final draft (if that's now let me know too), and I'll test it on my machine (and refactor the code organization if need be). Otherwise everything looks good and as long as it works, I'll pass a review. 

@albertz could you weigh in on a choice of name (both for the Github, package, and the command line tool)? My suggestion would be one of the following:

**Project:**
1. Scroll Wheel Accelerator
2. Scroll Accelerator

**Package and CLI**
1. `scroll-accelerator`
2. `scroll-wheel-accelerator`
3. `scroll-accel`
4. `scroll-wheel-accel`

Lmk what you guys think, this is nonexhaustive.

albertz commented:
I like `scroll-accelerator`. So that could be the name for the PIP package, for the GitHub repo, for the CLI binary. And `scroll_accelerator` for the Python package.

albertz commented:
> This change still misses the point of removing implementation from `__init__.py`

I agree that there should be no code in `__init__.py` (only imports maybe).

albertz commented:
> In my opinion package handling in Python is flawed by design: you shouldn't need virtual environments at all if you have a "package manager" (pip)

I personally often also use just one global environment for everything for simplicity. If sth broke with newer versions, I would just fix it. My mindset was that this is the best/easiest way to make sure that all the tools that I use always work correctly with the latest versions of all the libraries, as I always keep all libraries updated.

But I also see that many people don't like it when things break, and they want to have an isolated environment for every tool. And I think this is actually the more common case. So it's fine if the documentation / scripts assume this as the default.

Actually, I think newer Python versions don't allow for this `pip install --user` anymore, right? You will get this "error: externally-managed-environment" error (if it is not already some local/virtual env).

But it would be nice if any kind of usage (venv, uv, or no env) would work, if possible. We should not force sth to the user (like the usage of `uv`). But I think this is not done here, or is it? And the discussion about `__init__` or `__main__` is really orthogonal of this, or not?

eeriemyxi commented:
@albertz @valmikikothare Please have a look at https://github.com/eeriemyxi-contributions/mouse-scroll-wheel-acceleration-userspace/tree/feat-spread/src/mswaa and let me know if you like the new structure. You can suggest changes if you'd like. I'll push it to `main` when the changes are to everyone's liking.

I would've liked to avoid making this pull request from `main` but I'm not good with git's CLI and have lost important changes due to my ignorance in the past so I decided not to.

eeriemyxi commented:
> But I think this is not done here, or is it? And the discussion about __init__ or __main__ is really orthogonal of this, or not?

Yes, the code just executes an executable (in `install-daemon.py` script.) The user is free however they want to run the project through that executable. We can help them with instructions in the README instead.

albertz commented:
The `src/mswaa` dir should be renamed to `scroll_accelerator`. (So both `mswaa` -> `scroll_accelerator`, but also removing the extra `src/`.)

`impl.py` should be renamed to `scroll_accelerator.py`.

eeriemyxi commented:
I've renamed `impl.py` to `scroll_accelerator.py`. Rest of your requested changes can be done in another batch of commits that is focused on renaming the package.

@valmikikothare Let me know what you think of the new structure.

eeriemyxi commented:
> Please ping me when you have a final draft

I have more changes planned in the README as well as a few helper scripts to help with the installation. Once I'm done with those, it should be ready for a review. I'll then ping everyone relevant of course.

albertz commented:
> I've renamed `impl.py` to `scroll-accelerator.py`

But it's important to call it `scroll_accelerator.py`. `scroll-accelerator.py` is not valid.

eeriemyxi commented:
> But it's important to call it scroll_accelerator.py. scroll-accelerator.py is not valid.

Ah, typo in my comment. The actual file is correctly named.

albertz commented:
I would still also remove the `src/` prefix, and rename `mswaa` to `scroll_accelerator`.

eeriemyxi commented:
Yes, I plan to do them but doing that alone will break the package since the package name is used in `pyproject.toml` and likely other files. For now I'm focusing on the structure of `src/mswaa`, which will eventually be renamed to `scroll_accelerator`.

Hatchling, the build system, expects a `src` directory, so some configuration is also required in the `pyproject.toml` file for bare `scroll_accelerator` folder.

albertz commented:
Yes sure you also need to adapt pyproject.toml.

valmikikothare commented:
Sorry for the delay, been quite busy with work. Did a major overhaul of the project, you can see it in the `valmikikothare-feature` branch. I tried pushing to the branch associated with this pull request but I don't think I have permissions to (@eeriemyxi you might have to enable changes using the "“Allow edits from maintainers.” button). I welcome you guys to try it out, I think it is a much more polished package now

eeriemyxi commented:
Hi, I've been out of the loop from changes concerning these improvements for quite a long while at this point. I think it'd make more sense to make another pull request with the changes you've done and closing this one, since I don't really know if I can make good decisions at this point when I've forgotten major details of the implementation I did and also the project itself.

valmikikothare commented:
Sounds good, will do
